presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java

```diff
@@ -428,42 +447,71 @@ private PlanBuilder aggregate(PlanBuilder subPlan, QuerySpecification node)
         // 1. Pre-project all scalar inputs (arguments and non-trivial group by expressions)
         Set<Expression> groupByExpressions = ImmutableSet.copyOf(analysis.getGroupByExpressions(node));
 
-        ImmutableList.Builder<Expression> arguments = ImmutableList.builder();
-        analysis.getAggregates(node).stream()
+        List<Expression> aggregateArguments;
+        List<Expression> sortArguments;
+        List<Expression> filterArguments;
+
+        aggregateArguments = analysis.getAggregates(node).stream()
                 .map(FunctionCall::getArguments)
                 .flatMap(List::stream)
-                .forEach(arguments::add);
+                .collect(toImmutableList());
 
-        analysis.getAggregates(node).stream()
+        sortArguments = analysis.getAggregates(node).stream()
                 .map(FunctionCall::getOrderBy)
                 .filter(Optional::isPresent)
                 .map(Optional::get)
                 .map(OrderBy::getSortItems)
                 .flatMap(List::stream)
                 .map(SortItem::getSortKey)
-                .forEach(arguments::add);
+                .collect(toImmutableList());
 
         // filter expressions need to be projected first
-        analysis.getAggregates(node).stream()
+        filterArguments = analysis.getAggregates(node).stream()
                 .map(FunctionCall::getFilter)
                 .filter(Optional::isPresent)
                 .map(Optional::get)
-                .forEach(arguments::add);
+                .collect(toImmutableList());
+
+        Iterable<Expression> inputs = Iterables.concat(groupByExpressions, aggregateArguments, sortArguments, filterArguments);
+        ImmutableList<Expression> arguments = ImmutableList.copyOf(inputs);
 
-        Iterable<Expression> inputs = Iterables.concat(groupByExpressions, arguments.build());
         subPlan = handleSubqueries(subPlan, node, inputs);
 
-        if (!Iterables.isEmpty(inputs)) { // avoid an empty projection if the only aggregation is COUNT (which has no arguments)
+        Optional<List<FunctionCallWithFinalArguments>> functionCallWithFinalArgumentsOptional = Optional.empty();
```

Comment:
As discussed offline, rewriting the aggregate to ``CASE`` expression earlier than the place where it is done today will lead to losing original aggregates, and since ``aggregate`` will return ``PlanBuilder`` only with final translation, the ``project`` method called after the ``aggregate`` will project symbols from ``analysis``'s output expression, so there will be a dependency mismatch.